sdk/python/feast/infra/online_stores/postgres_online_store/postgres.py

```diff
@@ -456,6 +480,267 @@ def retrieve_online_documents(
+                    ORDER BY distance
+                    LIMIT {top_k}
+                    """
+                ).format(
+                    distance_metric_sql=sql.SQL(distance_metric_sql),
+                    table_name=sql.Identifier(table_name),
+                    top_k=sql.Literal(top_k),
+                )
+
+                cur.execute(query, (embedding, tsquery_str, string_fields, tsquery_str))
+                rows = cur.fetchall()
+
+            # Case 2: Vector Search Only
+            elif embedding is not None:
+                query = sql.SQL(
+                    """
+                    SELECT
+                        entity_key,
+                        feature_name,
+                        value,
+                        vector_value,
+                        vector_value {distance_metric_sql} %s::vector as distance,
+                        NULL as text_rank, -- Keep consistent columns
+                        event_ts,
+                        created_ts
+                    FROM {table_name}
+                    ORDER BY distance
+                    LIMIT {top_k}
+                    """
+                ).format(
+                    distance_metric_sql=sql.SQL(distance_metric_sql),
+                    table_name=sql.Identifier(table_name),
+                    top_k=sql.Literal(top_k),
+                )
+
+                cur.execute(query, (embedding,))
+                rows = cur.fetchall()
+
+            # Case 3: Text Search Only
+            elif query_string is not None and string_fields:
+                tsquery_str = " & ".join(query_string.split())
+                query = sql.SQL(
+                    """
+                    WITH text_matches AS (
+                        SELECT DISTINCT entity_key, ts_rank(to_tsvector('english', value_text), to_tsquery('english', %s)) as text_rank
+                        FROM {table_name}
+                        WHERE feature_name = ANY(%s) AND to_tsvector('english', value_text) @@ to_tsquery('english', %s)
+                        ORDER BY text_rank DESC
+                        LIMIT {top_k}
+                    )
+                    SELECT
+                        t1.entity_key,
+                        t1.feature_name,
+                        t1.value,
+                        t1.vector_value,
+                        NULL as distance,
+                        t2.text_rank,
+                        t1.event_ts,
+                        t1.created_ts
+                    FROM {table_name} t1
+                    INNER JOIN text_matches t2 ON t1.entity_key = t2.entity_key
+                    WHERE t1.feature_name = ANY(%s)
+                    ORDER BY t2.text_rank DESC
+                    """
+                ).format(
+                    table_name=sql.Identifier(table_name),
+                    top_k=sql.Literal(top_k),
+                )
+
+                cur.execute(
```

Comment:
cur.execute and cur.fetchall() is repeated in all conditions. 

if hybrid search, params = [embedding, tsquery_str, string_fields, tsquery_str]
if vector search, params = [embedding],
.....

cur.execute(query, params)
rows = cur.fetchall()